enclave/core/sgx/thread.c

```diff
@@ -167,65 +167,148 @@ typedef struct _oe_mutex_impl
     /* The thread that has locked this mutex */
     oe_sgx_td_t* owner;
 
+    /* The type of mutex (supported types: normal and recursive) */
+    uint32_t type;
+
     /* Queue of waiting threads (front holds the mutex) */
     Queue queue;
 } oe_mutex_impl_t;
 
+typedef struct _oe_mutexattr_impl
+{
+    uint32_t type;
+} oe_mutexattr_impl_t;
+
 OE_STATIC_ASSERT(sizeof(oe_mutex_impl_t) <= sizeof(oe_mutex_t));
+OE_STATIC_ASSERT(sizeof(oe_mutexattr_impl_t) <= sizeof(oe_mutexattr_t));
 
-oe_result_t oe_mutex_init(oe_mutex_t* mutex)
+static oe_result_t _validate_mutexattr_type(uint32_t type)
 {
-    oe_mutex_impl_t* m = (oe_mutex_impl_t*)mutex;
+    oe_result_t result = OE_FAILURE;
+
+    switch (type)
+    {
+        case OE_MUTEX_NORMAL:
+        case OE_MUTEX_RECURSIVE:
+        case OE_MUTEX_ERRORCHECK:
+            break;
+        default:
+            OE_RAISE(OE_INVALID_PARAMETER);
+    };
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_mutexattr_init(oe_mutexattr_t* attr)
+{
+    oe_mutexattr_impl_t* attr_impl = (oe_mutexattr_impl_t*)attr;
+    oe_result_t result = OE_FAILURE;
+
+    if (!attr_impl)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    attr_impl->type = OE_MUTEX_DEFAULT;
```

Comment:
To be consistent with oe_mutex_init, the entire oe_mutexattr_t struct needs to be set to zero. If we set only the `type` field, other portions of the struct could have random value.